python/cudf/cudf/core/index.py

```diff
@@ -1411,24 +1488,527 @@ def __cuda_array_interface__(self):
         return self._values.__cuda_array_interface__
 
 
+class Int8Index(GenericIndex):
+    """Immutable, ordered and sliceable sequence of integer labels.
+    The basic object storing row labels for all cuDF objects.
+    Int64Index is a special case of Index with purely integer labels.
+
+    Parameters:
+    -----------
+    data : array-like (1-dimensional)
+    dtype : NumPy dtype (default: int8)
+            If dtype is None, we find the dtype that best fits the data.
+    copy : bool
+        Make a copy of input data.
+    name : object
+        Name to be stored in the index.
+
+    Returns
+    -------
+    Index
+
+    Examples
+    --------
+    >>> import cudf
+    >>> cudf.Int64Index([1, 2, 3, 4], name="a")
+    Int64Index([1, 2, 3, 4], dtype='int8', name='a')
+
+    >>> cudf.Int64Index([1, 2, 3, 4], dtype="int16", name="a")
+    Int64Index([1, 2, 3, 4], dtype='int8', name='a')
+    """
+
+    def __new__(cls, data=None, dtype="int8", copy=False, name=None):
+
+        out = Frame().__new__(cls)
+
+        out._initialize_numeric(
+            data=data, dtype=dtype, expected_dtype="int8", copy=copy, name=name
+        )
+
+        return out
```

Comment:
Given this is mostly boilerplate repeated for a bunch of these classes, any thoughts on building a class factory method and using that instead to debloat the code?

Reply:
In all of these cases, why are we defining `__new_` for each of the derived types? Can we do something like this:

```python
class BaseNumericIndex(..):
    def __init__(self, data=None, dtype="int8", copy=False, name=None):
        ...

class Int8Index(BaseNumericIndex):
    def __init__(self, data, copy=False, name=None): # doesn't make sense to accept dtype here
        super.__init__(data, copy=copy, dtype="int8", name=name)  
     
```

Reply:
In that case mostly only we could eliminate two lines of code as documentation would be required all these different classes.

Reply:
Since the derived classes will differ from `BaseNumericIndex` only in their `dtype`, I think it'd be pretty reasonable to use something like [`copy_docstring`](https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/utils/docutils.py#L58) to copy the docstring of the base class to all of the derived classes

Reply:
@shwina That might work, but in that case I have to remove all the `__new__` cases from parents as well as they will get called before `__init__` might get called.But, that might work.

Reply:
`BaseNumericIndex` had to have `__new__` as upper classes are using `__new__`,  and `__new__` was required in the first place as we are creating objects in Index using `as_index.`

Reply:
Hmm - I still think we need to figure out a better approach here. It feels awkward that we're constructing a new `Frame` and starting from there. Typically `__new__` should call `super().__new__`. This becomes especially important when doing things like multiple inheritance. Will give this some thought.

Reply:
@rgsl888prabhu @kkraus14 Please feel free to merge this though. I'm happy to look into this in a follow-up.